theano/tensor/opt.py

```diff
@@ -2706,6 +2723,7 @@ def merge_two_slices(slice1, len1, slice2, len2):
             val = T.switch(T.lt(sl2, 0), - len1 - 1, val)
             if sl1.step:
                 val = T.switch(T.eq(sl1.step, 0), len1 + 1, val)
+            val = pre_greedy_local_optimizer(list_opt, val)
```

Comment:
This and the next one is important. It reapply what was done on line 2802 here. This make this optimization return a cleaned graph instead of making it generate ugly stuff that is cleaned up later. This is faster.